File: 2021/2/Main.hs
{-# LANGUAGE RecordWildCards #-}

import System.Environment (getArgs)
-- NOTE: would've been easier with Lenses

type Instruction = (String, Int)

getData :: String -> IO [Instruction]
getData fileName = do
  content <- readFile fileName
  pure $ map (readLine . words) $ lines content
  where
    readLine :: [String] -> Instruction
    readLine [inst, numStr] = (inst, read numStr)
    readLine _ = error "WTF?"

solve :: Position -> AimChoice -> [Instruction] -> Position
solve startPos ac =
  flip foldl startPos $ \pos (inst, num) -> case inst of
    "forward" -> moveForward num ac pos
    "up"      -> updateAimOrDepth num ac pos
    "down"    -> updateAimOrDepth (-num) ac pos
    _ -> undefined

makeScore :: Position -> Int
makeScore ps@Position{..} = _unX * _unZ

data Position = Position {
    _unX   :: Int
  , _unZ   :: Int
  , _unAim :: Int
} deriving (Show, Eq)

type PositionUpdater = Int -> AimChoice -> Position -> Position

moveForward :: PositionUpdater
moveForward num ac pos@Position{..} =
  if ac == WithoutAim
    then pos { _unX = _unX + num }
    else pos { _unX = _unX + num, _unZ = _unZ + (_unAim * num) }

updateAimOrDepth :: PositionUpdater
updateAimOrDepth num ac pos@Position{..} =
  if ac == WithoutAim
    then pos { _unZ = _unZ + num }
    else pos { _unAim = _unAim + num }

data AimChoice = WithAim | WithoutAim
  deriving(Eq)

main :: IO ()
main = do
  args <- getArgs
  input <- getData $ head args
  
  showScore $ solve' WithoutAim input
  showScore $ solve' WithAim input
  where
    start = Position 0 0 0
    solve' = solve start
    showScore = print . abs . makeScore

File: 2021/2/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func readFile(fileName string) []string {
	f, err := ioutil.ReadFile(fileName)
	check(err)

	return strings.Split(string(f), "\n")
}

type Position struct {
	X   int
	Z   int
	Aim int
}

func (p *Position) MoveForward(unit int, toChangeAim bool) {
	p.X += unit
	if toChangeAim {
		p.Z += (p.Aim * unit)
	}
}

func (p *Position) MoveDepth(unit int, toChangeAim bool) {
	if !toChangeAim {
		p.Z += unit
		return
	}
	p.moveAim(unit)
}

func (p *Position) moveAim(unit int) {
	p.Aim += unit
}

func (p Position) Result() int {
	return int(math.Abs(float64(p.X * p.Z)))
}

func solve(insts []string, toChangeAim bool) {
	p := Position{X: 0, Z: 0, Aim: 0}

	for _, instStr := range insts {
		is := strings.Split(instStr, " ")
		inst := is[0]
		num, err := strconv.Atoi(is[1])
		check(err)

		if inst == "forward" {
			p.MoveForward(num, toChangeAim)
		} else if inst == "up" {
			p.MoveDepth(num, toChangeAim)
		} else if inst == "down" {
			p.MoveDepth(-num, toChangeAim)
		}
	}

	fmt.Println(p.Result())
}

func main() {
	fileName := os.Args[1]
	lines := readFile(fileName)

	solve(lines, false)
	solve(lines, true)
}
